rust/kernel/platdev.rs

```diff
@@ -30,13 +30,31 @@ pub struct Registration {
 // (it is fine for multiple threads to have a shared reference to it).
 unsafe impl Sync for Registration {}
 
+extern "C" {
+    #[allow(improper_ctypes)]
```

Comment:
What's the improper ctype here?

Reply:
Without it, I get the following warning:
```rust
  RUSTC L rust/kernel.o
warning: `extern` block uses type `arch_spinlock_t`, which is not FFI-safe
    --> rust/kernel/platdev.rs:36:15
     |
36   |         pdev: *const bindings::platform_device,
     |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not FFI-safe
     |
     = note: `#[warn(improper_ctypes)]` on by default
     = help: consider adding a member to this struct
     = note: this struct has no fields
note: the type is defined here
    --> /home/sva/repos/kernel-arcx/rust/bindings_generated.rs:7331:1
     |
7331 | pub struct arch_spinlock_t {}
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: 1 warning emitted
```

Reply:
What type is using `arch_spinlock_t`? Is it `raw_spinlock`? You can add the type to the opaque list: https://github.com/Rust-for-Linux/linux/blob/d8c6b9ce7fb78375460422ba15d763c22f70d0af/rust/bindgen_parameters#L10.

Reply:
Hi Gary, thanks for the help! I suspect that on arm 32-bit, `arch_spinlock_t` is just its own type, and not a typedef. So should I add it to `--opaque-type`?
https://github.com/Rust-for-Linux/linux/blob/d8c6b9ce7fb78375460422ba15d763c22f70d0af/arch/arm/include/asm/spinlock_types.h#L11-L24

Reply:
Ok, when I add `arch_spinlock_t` to `--opaque-type`, I seem to enter some kind of rabbit-hole where different structures trigger the warning, and when you add those to `--opaque-type`, there are more structures that trigger it. So far I have:
```
--opaque-type arch_spinlock_t
--opaque-type lock_class_key
--opaque-type arch_rwlock_t
--opaque-type vm_userfaultfd_ctx
```
and still the warnings keep coming.

Reply:
Setting `arch_spinlock_t` itself to opaque type just propagate the issue one struct up :)

Do you need to peek into fields of `platform_device`? If not, just setting that to opaque type will be sufficient.

Reply:
Lol good point... will do. Thanks !

Reply:
But the peeking is just a hack, and will disappear once we proceed. I'll make a mental note to add `platform_device` to `--opaque-type` later, and remove the `#[allow(improper_ctypes)]`.